### DISCUSSION TITLE
Lab3 registers usage problem

### DISCUSSION THREAD

**Discussion by JackeyUi** [bol-edu/HLS-SOC-Discussions#86] [Q&A]
關於Lab3的spec中，FIR內部register的使用還有些不清楚的地方：
1. ap_start, ap_done, ap_idle跟 data_length 都是存在FIR裡面嗎？如果是的話那 ap ctrl 的那三個訊號是直接用3 bits的reg去存就好，還是要開個 32 bits的 reg去存呢？
2. 只用一個乘法器跟一個加法器時，FIR每一步的運算式為 : temp'  =  coef[i] * data[t-i] + temp
而 coef 存在 tab_RAM， temp(計算過程的中間值)存在 shift_RAM，那FIR內必須得有一個data 的register 來暫存從 ss 端收到的值，想請問這是被允許的嗎？因為我想了很久，計算要成功，除了在FIR 開個 32 bits 的registers 去存data 以外，唯一的辦法就是在ss_tready 拉高之前先去 acess ss_tdata，然後等算完之後再拉高 ss_tready ，下一筆 data 再進來。但我的理解這樣應該是違反 AXI4-Stream Transfer Protocol 的，所以想問該怎麼辦？或是有甚麼辦法是我沒想到的？

**Comment by ZheChen-Bill** [comment]
1.我們是使用類似於lab2的address map，在lab2中的定義如下。
![圖片](https://github.com/bol-edu/HLS-SOC-Discussions/assets/88698677/656ad6d5-6a61-4e18-9afa-ef27f3c81b7e)
2.因為我自己在設計FIR時，發現Coef和Data會大量的重複使用，因此將其儲存在BRAM中，但應該還有很多方法可以進行設計，關於你的temp或許可以查詢一下FF是如何運作的。

**Reply by JackeyUi** [reply]
2. 我不太懂你的意思。如果你是將coef和data存在bram，那你的temp不就該在FIR裡面開個暫存器去存它嗎？

**Reply by ZheChen-Bill** [reply]
是的。

**Reply by BarryJu1999** [reply]
不好意思打擾您，我發現我沒很懂lab 2中的內容還是想請教關於第一點的內容，

這個address map實際代表的是指要讀取的資料為何，
但其中資料的儲存方式其實並無限制，
像是ap_start、ap_done、ap_idle其實可以使用3bit reg儲存(只有用到這三個)，
另外像是tap parameter或是data_num(reserved)因lab3要求係數要使用bram，而data_num並無限制，
至於address就要自己在接收後轉換，如轉換到bram中的A[0:10]中的address，

以上是我的理解我不太確定我想的對不對，希望能再請教您，抱歉打擾您了。

**Reply by ghost** [reply]
我也想問，是不是data_num不用存在bram，因為如果必須存的話，變成必須使用bram12才夠
